Add PAFTP to Wi-Fi PAF commissioning

crlonxp opened the pull request:
* Change the default setting of WiFiPAF as enabled if running on Linux system
* Change to use C++ casting
* Move the paf cancel function from destructor to shutdown
* Add PAFTP

### Testing
* Environment: 
    AP: Asus RT-N66U, running at chnl#1
    ctrl: Rpi + NetGear, Inc. A6210
    dut#1: rpi + Linksys AE6000
    dut#2: imx93+iw612
* dut: 
    $ sudo ./chip-all-clusters-app --wifi --wifipaf freq_list=2412 
* ctrl: Connect with AP before running the test
    $ sudo ./chip-tool pairing code-wifi 1 n_m_2g nxp12345 MT:-24J0SGJ10KA0648G00 --freq 2412
    $ sudo ./chip-tool pairing wifipaf-wifi 1 n_m_2g nxp12345 20202021 3840 --freq 2412

crlonxp committed:
* Fix the dependency error
* Change the default setting of WiFiPAF as enabled if running on Linux system
* Change to use C++ casting
* Move the paf cancel function from desctructor to shutdown

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

crlonxp committed:
* Add the missing files

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

crlonxp committed:
Not failed to run immediately if wpa_supplicant is not ready

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
* Fix the error found by the building script

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

andy31415 commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -1823,6 +1820,18 @@ void DeviceCommissioner::CloseBleConnection()
+{
+    Transport::WiFiPAFBase & transport =
+        std::get<Transport::WiFiPAF<1>>(mSystemState->TransportMgr()->GetTransport().GetTransports());
+    if (!transport.IsWiFiPAFLayerTransportSetToSelf())
```
Is this conditional needed? Why would the transport not include this internally?

crlonxp replied:
I will change it

andy31415 commented on src/platform/device.gni:
```diff
@@ -25,7 +25,7 @@ declare_args() {
 
   # Include wifi-paf to commission the device or not
-  chip_device_config_enable_wifipaf = false
+  chip_device_config_enable_wifipaf = current_os == "linux"
```
why did we change this default? Could you add a comment here on sideffects?

I am not familiar with PAF - do we expect any linux device with a wifi card to support this or do you require special hardware/drivers? Is this expected to be transparent for devices that do not support it?

crlonxp replied:
This is as discussed on Slack to enable PAF in TH we need to have it enable by default.
Is this wrong understanding?

andy31415 commented on src/platform/device.gni:
```diff
@@ -124,6 +124,10 @@ declare_args() {
 }
 
+if (chip_enable_wifi == false) {
+  chip_device_config_enable_wifipaf = false
```
you are resetting this later ... why not include this in the original condition like:

```
chip_device_config_enable_wifipaf = chip_enable_wifi && (current_os == "linux") 
```

crlonxp replied:
I will change it

andy31415 commented on src/transport/raw/WiFiPAF.h:
```diff
@@ -24,6 +24,7 @@
 
 #pragma once
 
+#include "wifipaf/WiFiPAFLayer.h"
```
use `<>` for includes

crlonxp replied:
I will change it

andy31415 commented on src/transport/raw/WiFiPAF.h:
```diff
@@ -69,22 +54,20 @@ class DLL_EXPORT WiFiPAFBase : public Base
-    State mState = State::kNotReady;
+
+    WiFiPAF::WiFiPAFLayer * mWiFiPAFLayer = nullptr; ///< Associated wifipaf layer
+    // WiFiPAF::WiFiPAFLayer::State mState = WiFiPAF::WiFiPAFLayer::State::kNotReady;
```
remove commented out code

crlonxp replied:
I will remove it

woody-apple reviewed (changes requested):
An exception is required for this feature, we cannot enable this by default in the SDK at this point.

crlonxp committed:
Change by followign the review's comment

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

restyled-commits committed:
Restyled by gn

restyled-commits committed:
Restyled by clang-format

woody-apple commented on src/platform/device.gni:
```diff
@@ -119,6 +116,13 @@ declare_args() {
   }
 }
 
+declare_args() {
+  # Include wifi-paf to commission the device or not
+  # This is a feature of Wi-Fi spec that it can be enabled if wifi is enabled
+  # and the supplicant can support.
+  chip_device_config_enable_wifipaf = chip_enable_wifi && current_os == "linux"
+}
+
```
```suggestion
```

woody-apple replied:
Please remove this per the TT discussion today, I'll help with a separate workflow to test this.

woody-apple commented on src/platform/device.gni:
```diff
@@ -23,9 +23,6 @@ declare_args() {
 
   # Substitute fake platform when building with chip_device_platform=auto.
   chip_fake_platform = false
-
-  # Include wifi-paf to commission the device or not
-  chip_device_config_enable_wifipaf = false
```
Let's add this back in

crlonxp replied:
May we change it as: 
chip_device_config_enable_wifipaf = false && chip_enable_wifi
Since it can be enabled only while wifi is enabled.

crlonxp committed:
Disable WiFiPAF by default

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Separate the dbus parameters to have their own type, instead of using the string

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

restyled-commits committed:
Restyled by clang-format

crlonxp committed:
Change the dbus calling method by following the new defined api of wpa_supplicant

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Fix a coredump if chip_device_config_enable_wifipaf is set, but --wifipaf is not selected

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Restart DnssdServer on connectivity change, not only on ipv4, but also on ipv6

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Fix the compile error after rebasing to master branch

crlonxp committed:
* Block the multiple indication of discovery-result
* Fix the problems in NANCancelPublish by using the incorrect type and id

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

andy31415 commented on src/platform/Linux/BUILD.gn:
```diff
@@ -154,6 +154,10 @@ static_library("Linux") {
   }
+
+  if (chip_device_config_enable_wifipaf) {
+    public_deps += [ "${chip_root}/src/wifipaf:wifipaf" ]
```
```suggestion
    public_deps += [ "${chip_root}/src/wifipaf" ]
```

andy31415 commented on src/platform/device.gni:
```diff
@@ -119,6 +116,13 @@ declare_args() {
+  # Include wifi-paf to commission the device or not
+  # This is a feature of Wi-Fi spec that it can be enabled if wifi is enabled
+  # and the supplicant can support.
+  chip_device_config_enable_wifipaf = chip_enable_wifi && false
```
&& false means false.

Maybe you can assert that if one is true the other one is as well instead.

crlonxp replied:
This is to be asked to disable it for now. The correct one should be:
	chip_device_config_enable_wifipaf = chip_enable_wifi && (chip_device_platform == "linux")

PAF should be enabled by default if:
	- Wi-Fi is enabled && 
	- It's in Linux (Since only wpa_supplicant has this feature for now)

andy31415 commented on src/lib/BUILD.gn:
```diff
@@ -34,6 +34,10 @@ source_set("without-logging") {
     "${chip_root}/src/transport",
   ]
 
+  if (chip_device_config_enable_wifipaf == true) {
```
```suggestion
  if (chip_device_config_enable_wifipaf) {
```

andy31415 commented on src/include/platform/ConnectivityManager.h:
```diff
@@ -289,12 +289,15 @@ struct ConnectivityManager::SEDIntervalsConfig
 };
 
 #if CHIP_DEVICE_CONFIG_ENABLE_WIFIPAF
+#define NAN_FREQ_LIST_ALL 0xff
```
use `constexpr uint8_t kNaNFreqListAll` (or whatever type or name is apropriate instead) to get type safety as well.

crlonxp replied:
I will fix it

andy31415 commented on src/include/platform/CHIPDeviceConfig.h:
```diff
@@ -447,6 +447,60 @@
+ * The max amount of time (in seconds) the chip controller will discovery Wi-Fi PAF
+ */
+#ifndef CHIP_DEVICE_CONFIG_WIFIPAF_DISCOVERY_TIMEOUT
+#define CHIP_DEVICE_CONFIG_WIFIPAF_DISCOVERY_TIMEOUT (15 * 60)
```
please add unit of measurement suffix to all these constants. for seconds it should probably be `_SEC`

crlonxp replied:
I will fix it

andy31415 commented on examples/platform/linux/AppMain.cpp:
```diff
@@ -369,6 +369,71 @@ class SampleTestEventTriggerHandler : public TestEventTriggerHandler
 };
 
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFIPAF
+void int_array_add_unique(uint16_t ** res, uint16_t a)
```
Add a comment on what this method does and usage examples.

crlonxp replied:
I will change it

andy31415 commented on examples/platform/linux/AppMain.cpp:
```diff
@@ -369,6 +369,71 @@ class SampleTestEventTriggerHandler : public TestEventTriggerHandler
+void int_array_add_unique(uint16_t ** res, uint16_t a)
+{
+    size_t reslen, max_size;
+    uint16_t * n;
```
move this closer to usage instead of at a top. Also give it a better name than `n`. What does it mean.

TLDR: I cannot tell at a glance what this method is doing and what its logic or variables mean. Please make this readable.

crlonxp replied:
This is actually extracted from wpa_supplicant source code which processes the similar function. I will rewrite it

andy31415 commented on examples/platform/linux/AppMain.cpp:
```diff
@@ -369,6 +369,71 @@ class SampleTestEventTriggerHandler : public TestEventTriggerHandler
+    *res          = n;
+}
+
+static uint16_t WiFiPAFGet_FreqList(char * ArgStrn, uint16_t ** pfreq_list)
```
Please also improve readability on this method to:
  - describe what it does and how (what do the arguments mean, how are they transformed, usage examples
  - rename some variables (what is `len` for example)?
  - is `freq_list` needed? we seem to use it just for a get and to later place it in pfreq_list. If needed this should be moved to decrease its scope.
  - Can we somehow make it so that ArgStrn can be const? generally this seems the intent and we use non-const so that strtok works.

crlonxp replied:
I will change it

andy31415 commented on examples/platform/linux/AppMain.cpp:
```diff
@@ -484,16 +549,21 @@ int ChipLinuxAppInit(int argc, char * const argv[], OptionSet * customOptions,
-            args.enable  = LinuxDeviceOptions::GetInstance().mWiFiPAF;
-            args.ExtCmds = LinuxDeviceOptions::GetInstance().mWiFiPAFExtCmds;
+            args.enable        = LinuxDeviceOptions::GetInstance().mWiFiPAF;
+            args.freq_list_len = WiFiPAFGet_FreqList((char *) LinuxDeviceOptions::GetInstance().mWiFiPAFExtCmds, &args.pfreq_list);
```
cast is suspicious. You should use C++-style casts, and then it is a `const_cast` and that is a code-smell. Fix up WiFiPAFGet_FreqList to accept const char instead.

crlonxp replied:
I will fix it

andy31415 commented on examples/platform/linux/AppMain.cpp:
```diff
@@ -484,16 +549,21 @@ int ChipLinuxAppInit(int argc, char * const argv[], OptionSet * customOptions,
+            args.enable        = LinuxDeviceOptions::GetInstance().mWiFiPAF;
+            args.freq_list_len = WiFiPAFGet_FreqList((char *) LinuxDeviceOptions::GetInstance().mWiFiPAFExtCmds, &args.pfreq_list);
             DeviceLayer::ConnectivityMgr().SetWiFiPAFAdvertisingEnabled(args);
+            if ((args.freq_list_len > 0) && (args.freq_list_len != NAN_FREQ_LIST_ALL))
```
explain this code in a comment? why are we freeng all except the `all` constant?

Why can't this be automatic (like having the ARGS handle memory managment, not the caller)?

crlonxp replied:
I will fix it

crlonxp committed:
Update src/lib/BUILD.gn

Co-authored-by: Andrei Litvin <andy314@gmail.com>

crlonxp committed:
Update src/platform/Linux/BUILD.gn

Co-authored-by: Andrei Litvin <andy314@gmail.com>

crlonxp committed:
* Add the unit to the time-out const
* Refine the function to transfer the freq_list from user to internal structure

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add "freq" option of chip-tool to pass down the frequency of the connected ap

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add a vector to management the PAF sessions
Miscellaneous bugs fixing

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Fix the errors found from the ci/cv

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
* Bug fix: Add the PAF session if using code-wifi
* Remove unused code

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add to cancel PAF subscribe-session if user inputs ctrl+C to stop chip-tool manually

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add PAFTP implementation

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add the missing .h to BUILD.gn for code-lints

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Fix the clang error

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add to close PAF session while closing EndPoint module

crlonxp committed:
Add the re-order processing right after receiving the PAF packets
Cancel the publish if the previous fabric information has existed

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add the missing change of PAF in rebasing to TOT

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

tcarmelveilleux commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -710,6 +707,23 @@ CHIP_ERROR DeviceCommissioner::EstablishPASEConnection(NodeId remoteDeviceId, co
                                        resolutionData);
 }
 
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFIPAF
+static void StopSignalHandler(int signum)
+{
+    WiFiPAF::WiFiPAFLayer::GetWiFiPAFLayer()->Shutdown([](uint32_t id, WiFiPAF::WiFiPafRole role) {
+        switch (role)
+        {
+        case WiFiPAF::WiFiPafRole::kWiFiPafRole_Publisher:
+            DeviceLayer::ConnectivityMgr().WiFiPAFCancelPublish(id);
+            break;
+        case WiFiPAF::WiFiPafRole::kWiFiPafRole_Subscriber:
+            DeviceLayer::ConnectivityMgr().WiFiPAFCancelSubscribe(id);
+            break;
+        }
+    });
+}
```
Signal handling not allowed directly in portable code. The signal handling must be in platform specific code and code here must remain fully portable.

crlonxp replied:
I will move them to src/platform/Linux/ConnectivityManagerImpl.cpp

tcarmelveilleux commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -831,17 +845,26 @@ CHIP_ERROR DeviceCommissioner::EstablishPASEConnection(NodeId remoteDeviceId, Re
+            const SetupDiscriminator connDiscriminator(params.GetSetupDiscriminator().value());
+            uint16_t discriminator =
+                connDiscriminator.IsShortDiscriminator() ? connDiscriminator.GetShortValue() : connDiscriminator.GetLongValue();
+            DeviceLayer::ConnectivityMgr().GetWiFiPAF()->AddPafSession(nodeId, discriminator);
+            DeviceLayer::ConnectivityMgr().WiFiPAFSubscribe(params.GetSetupDiscriminator().value(), reinterpret_cast<void *>(this),
+                                                            OnWiFiPAFSubscribeComplete, OnWiFiPAFSubscribeError);
```
This must be conditional on Wi-Fi PAF enabled.

tcarmelveilleux reviewed (dismissed):
- [ ] Remove any mention of signal handling in portable code.
- [ ] Add testing for the feature. There are zero tests in this PR and it adds 4000+ lines of code.

crlonxp committed:
Move the signal handling function to Linux connectivity manager

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -831,17 +845,26 @@ CHIP_ERROR DeviceCommissioner::EstablishPASEConnection(NodeId remoteDeviceId, Re
+                                                            OnWiFiPAFSubscribeComplete, OnWiFiPAFSubscribeError);
+            signal(SIGINT, StopSignalHandler);
+            signal(SIGTERM, StopSignalHandler);
+
```
I will move them to src/platform/Linux/ConnectivityManagerImpl.cpp

crlonxp commented on src/controller/SetUpCodePairer.cpp:
```diff
@@ -266,8 +259,13 @@ CHIP_ERROR SetUpCodePairer::StartDiscoverOverWiFiPAF(SetupPayload & payload)
+    const SetupDiscriminator connDiscriminator(payload.discriminator);
+    uint16_t discriminator =
+        connDiscriminator.IsShortDiscriminator() ? connDiscriminator.GetShortValue() : connDiscriminator.GetLongValue();
+    DeviceLayer::ConnectivityMgr().GetWiFiPAF()->AddPafSession(mRemoteId, discriminator);
+    CHIP_ERROR err = DeviceLayer::ConnectivityMgr().WiFiPAFSubscribe(payload.discriminator, (void *) this,
+                                                                     OnWiFiPAFSubscribeComplete, OnWiFiPAFSubscribeError);
```
This is within CHIP_DEVICE_CONFIG_ENABLE_WIFIPAF compile flag.
Could you suggest me what should I do? Thank you

jrrech commented:
@crlonxp I'm trying to pair a device over wifipaf with your changes and running into `WiFi-PAF: DiscoveryResult, reentrance` and ultimately `PASESession timed out` errors.

May I ask which version of `wpa_supplicant` you're running there? Thanks!

crlonxp committed:
Fix the error after rebase to TOT of the master branch

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
* Fine tune the paf-rx path by adding the return value
* Remove the unused error code

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add unit-test of WiFiPAF

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

tcarmelveilleux commented on scripts/tools/check_includes_config.py:
```diff
@@ -188,7 +188,7 @@
     'src/controller/CHIPDeviceController.cpp': {'string'},
     'src/qrcodetool/setup_payload_commands.cpp': {'string'},
     'src/access/AccessRestrictionProvider.h': {'vector', 'map'},
-
```
Is this to be used by the device side? If so, this should not be overridden without justification.

crlonxp replied:
I will remove it.

crlonxp committed:
No using "<vector>" to fix the error from check_include_config.py
Change the management of PAFInfo items

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

tcarmelveilleux commented on src/platform/Linux/ConnectivityManagerImpl.cpp:
```diff
@@ -938,7 +941,8 @@ CHIP_ERROR ConnectivityManagerImpl::_WiFiPAFPublish(ConnectivityManager::WiFiPAF
     ChipLogProgress(DeviceLayer, "WiFi-PAF: publish_id: %u ! ", publish_id);
-    _GetWiFiPAF()->AddPafSession(publish_id);
+    WiFiPAFSession sessionInfo = { .id = publish_id };
+    ReturnErrorOnFailure(_GetWiFiPAF()->AddPafSession(PafInfoAccess::kAccSessionId, sessionInfo));
```
Please do not use symbols with leading underscores. Singletons should look like `WiFiPAF()::GetInstance()` and they should return references not pointers.

crlonxp replied:
I will change it.

tcarmelveilleux commented on src/wifipaf/WiFiPAFLayer.cpp:
```diff
@@ -384,84 +392,143 @@ WiFiPAFLayer::GetHighestSupportedProtocolVersion(const PAFTransportCapabilitiesR
-    PafInfoVect.push_back({ .id = id });
-    ChipLogProgress(WiFiPAF, "WiFiPAF: Add session with id: %u, total %lu sessions", id, PafInfoVect.size());
+    memset(&SessionInfo, 0, sizeof(WiFiPAFSession));
+    SessionInfo.id            = UINT32_MAX;
```
Should this be INVALID_PAF_SESSION_ID?

crlonxp replied:
It's confusing. I will change it

tcarmelveilleux commented on src/wifipaf/WiFiPAFLayer.cpp:
```diff
@@ -384,84 +392,143 @@ WiFiPAFLayer::GetHighestSupportedProtocolVersion(const PAFTransportCapabilitiesR
+        case PafInfoAccess::kAccNodeInfo:
+            if (pPafSession->nodeId == SessionInfo.nodeId)
+            {
+                assert(pPafSession->discriminator == SessionInfo.discriminator);
```
Direct asserts are forbidden. Please use `VerifyOrDie`

crlonxp replied:
I will change it

tcarmelveilleux commented on src/wifipaf/WiFiPAFLayer.cpp:
```diff
@@ -384,84 +392,143 @@ WiFiPAFLayer::GetHighestSupportedProtocolVersion(const PAFTransportCapabilitiesR
+        default:
+            return CHIP_ERROR_NOT_IMPLEMENTED;
+        };
+        if ((pPafSession->id == UINT32_MAX) && (pPafSession->nodeId == UINT32_MAX) && (pPafSession->discriminator == UINT16_MAX))
```
Please use curly braces after this if

crlonxp replied:
I will add it

tcarmelveilleux commented on src/wifipaf/tests/TestWiFiPAFLayer.cpp:
```diff
@@ -114,25 +114,31 @@ TEST_F(TestWiFiPAFLayer, CheckWiFiPAFTransportCapabilitiesResponseMessage)
+    EXPECT_EQ(pPafSessionInfo_nodeid->discriminator, sessionInfo.discriminator);
+
+    // Get the session info by giving the discriminator
+    sessionInfo.discriminator = 0xF00;
```
- Please test for search of a session that does not exist
- Please test for the situation where there are > 1 PAF session and you look for a given discriminator (e.g. both first and second), and make sure you don't find the other one.

crlonxp replied:
Yes, I will add this and more cases to improve the coverage.

crlonxp commented on src/wifipaf/WiFiPAFLayer.h:
```diff
@@ -182,15 +188,14 @@ class DLL_EXPORT WiFiPAFLayer
 private:
-    std::vector<WiFiPAFSession> PafInfoVect;
+    void InitialPafInfo();
+    void cleanPafInfo(WiFiPAFSession & SessionInfo);
```
I will change it

crlonxp committed:
Fix the error that PAF can't start multiple times in using chip-tool interactive mode
Fix the code style problems

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Fix the crash error while running the unit-test

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add more unit-test cases

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

tcarmelveilleux commented on src/wifipaf/tests/TestWiFiPAFTP.cpp:
```diff
@@ -0,0 +1,180 @@
+
+TEST_F(TestWiFiPAFTP, CheckSendMultiple)
+{
+    auto buf = System::PacketBufferHandle::New(500);
+    buf->SetDataLength(500);
```
This buffer is random data from the stack, so when you allocate, please clear them otherwise some static analyzers may complain that data is used uninitialized.

crlonxp replied:
I will add to clear the buffer. Thanks.

crlonxp committed:
Clear the buffer after allocation in the test module
Declare test class as the friend class to access the private functions for testing

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

tcarmelveilleux commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1273 @@
+
+CHIP_ERROR WiFiPAFEndPoint::_Receive(PacketBufferHandle && data)
+{
+    ChipLogDebugWiFiPAFEndPoint(WiFiPAF, "+++++++++++++++++++++ entered receive");
```
Please remove this level of verbosity.

crlonxp replied:
Yes, I will

tcarmelveilleux commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1273 @@
+    return err;
+}
+
+CHIP_ERROR WiFiPAFEndPoint::_Receive(PacketBufferHandle && data)
```
For all methods, please remove `_` for private methods, it's not needed to have previx.

crlonxp replied:
Yes, I will change the name

tcarmelveilleux commented on src/wifipaf/tests/TestWiFiPAFLayer.cpp:
```diff
@@ -0,0 +1,154 @@
+namespace chip {
+namespace WiFiPAF {
+
+class TestWiFiPAFLayer : public WiFiPAFLayer, private WiFiPAFLayerDelegate, public ::testing::Test
```
Furthermore, please add cases which cover incorrectly formatted packets, payloads, and reassembly of multiple segments.

crlonxp replied:
Yes, I have add some cases for re-ordering packets and will keep adding more incorrect formatted packets.

crlonxp replied:
The coverage now is:
	Line_coverage: 76.5%,
	Function_coverage: 91.4%

crlonxp committed:
Remove "_" for private method
Remove some verbose messages which are from initial debugging purpose

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add more unit test cases of wifipaf

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

crlonxp committed:
Add a new unit-test case: send a long packet whose size > mtu

Signed-off-by: Lo,Chin-Ran <chin-ran.lo@nxp.com>

bzbarsky-apple commented on src/transport/raw/WiFiPAF.h:
```diff
@@ -1,6 +1,6 @@
 /*
  *
- *    Copyright (c) 2024 Project CHIP Authors
+ *    Copyright (c) 2025 Project CHIP Authors
```
Again, why is this nuking the old copyright date?

crlonxp replied:
I will change it

bzbarsky-apple commented on src/wifipaf/WiFiPAFConfig.h:
```diff
@@ -0,0 +1,142 @@
+ *    simultaneous PAF connections the application will establish).
+ */
+#ifndef WIFIPAF_LAYER_NUM_PAF_ENDPOINTS
+#define WIFIPAF_LAYER_NUM_PAF_ENDPOINTS 2
```
OK, now I really don't get it.  If we support > 1 connection, how does any of the "shut down everything" logic I asked question about above make any sense?

bzbarsky-apple commented on src/wifipaf/WiFiPAFConfig.h:
```diff
@@ -0,0 +1,142 @@
+ *    of ONE-HALF the total number of slots or buffers reserved. The PAFTP layer reserves all of these buffers for its
+ *    own use - one half for incoming writes or indications, and the other half for incoming confirmations.
+ *
+ *    This value must be greater than 1, or race condition avoidance logic will prevent send the on remote device. This
```
What does "send the on remote device" mean?

crlonxp replied:
The description should be removed. I will change it.

bzbarsky-apple commented on src/wifipaf/WiFiPAFConfig.h:
```diff
@@ -0,0 +1,142 @@
+ *
+ *    This value must be greater than 1, or race condition avoidance logic will prevent send the on remote device. This
+ *    logic prevents any send with no piggybacked ack when the receiver's window has only 1 slot open. Without this
+ *    logic, simultaneous data transmissions could fill both receiver's windows, leaving no room for the acks required
```
"both receivers'"

crlonxp replied:
The description was inherited from the bletp. I will remove them and reference the detailed information on the specification.

bzbarsky-apple commented on src/wifipaf/WiFiPAFConfig.h:
```diff
@@ -0,0 +1,142 @@
+#endif
+
+#if (PAF_MAX_RECEIVE_WINDOW_SIZE < 3)
+#error "PAF_MAX_RECEIVE_WINDOW_SIZE must be greater than 2 for PAF transport protocol stability."
```
The comments above said "greater than 1", so where does "greater than 2" come from?

crlonxp replied:
It's to add one more spare space for not sending ack so often. But, yes it's not error to set it "2". I will change it

bzbarsky-apple commented on src/wifipaf/WiFiPAFConfig.h:
```diff
@@ -0,0 +1,142 @@
+ *    logic, simultaneous data transmissions could fill both receiver's windows, leaving no room for the acks required
+ *    to re-open them. Both senders would wedge, and the PAFTP connection would stall.
+ *
+ *    This value must also exceed (WIFIPAF_CONFIG_IMMEDIATE_ACK_WINDOW_THRESHOLD + 1), or ***immediate*** stand-alone
```
What's WIFIPAF_CONFIG_IMMEDIATE_ACK_WINDOW_THRESHOLD?  It's some sort of internal macro; how would someone configuring these bits know what it is?  It should at least be defined here, with relevant static asserts about the receive window size exceeding it, etc.

crlonxp replied:
I will change it. Thank you for the suggestion

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR WiFiPAFEndPoint::SendCharacteristic(PacketBufferHandle && buf)
```
Is it called a "characteristic" for PAF?  BLE has "characteristics", sure, but I find nothing called "characteristic" for PAF....

crlonxp replied:
It should be the follow-up packets. I will change it

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+    uint8_t SnOffset = 0;
+    SequenceNumber_t * pSn;
+    err = reader.Read8(rx_flags.RawStorage()).StatusCode();
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kHankshake))
```
What's a "Hankshake"?

crlonxp replied:
Oh, sorry. I will fix it.

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+    err = reader.Read8(rx_flags.RawStorage()).StatusCode();
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kHankshake))
+    {
+        // Handkshake message => No ack/sn
```
Or a "Handkshake"?

crlonxp replied:
yes

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+        SnOffset += kTransferProtocolAckSize;
+    }
+    pSn    = pHead + SnOffset;
+    seqNum = *pSn;
```
How do you know you didn't run off the end of the buffer here?  This is completely ignoring the actual length of the data, no?

crlonxp replied:
Yes. the checking is required

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+    SuccessOrExit(err);
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kHankshake))
+    {
+        // Handkshake message => No ack/sn
```
Again, please fix the spelling....

crlonxp replied:
Yes, I will

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+        pAct = pHead + kTransferProtocolHeaderFlagsSize;
+        SnOffset += kTransferProtocolAckSize;
+    }
+    pSn = pHead + SnOffset;
```
How do you `pSn` is not pointing off the end of the buffer?

crlonxp replied:
The checking is required. I will add it

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+
+    ChipLogDebugBufferWiFiPAFEndPoint(WiFiPAF, buf);
+    Encoding::LittleEndian::Reader reader(buf->Start(), buf->DataLength());
+    DebugPktAckSn(PktDirect_t::kTx, reader, buf->Start());
```
I really don't get it.  If you have a reader on the buffer, why are you _also_ manually trying to mess with the buffer pointer?

crlonxp replied:
I will change it

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+SequenceNumber_t WiFiPAFEndPoint::AdjustRemoteReceiveWindow(SequenceNumber_t lastReceivedAck, SequenceNumber_t maxRemoteWindowSize,
+                                                            SequenceNumber_t newestUnackedSentSeqNum)
+{
+    // Assumption: SequenceNumber_t is uint8_t.
```
Should be static_asserted.

crlonxp replied:
I will add it

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+        SnOffset += kTransferProtocolAckSize;
+    }
+    pSn    = pHead + SnOffset;
+    seqNum = *pSn;
```
How do you know `*pSn` is in native endianness?  If it's known to be one-byte, that should be static_asserted here where you assume it.

crlonxp replied:
The type of seqNum is SequenceNumber_t which is one-byte. I will add the static_assert

bzbarsky-apple commented on src/wifipaf/WiFiPAFEndPoint.cpp:
```diff
@@ -0,0 +1,1255 @@
+    }
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kFragmentAck)) // Has ack
+    {
+        pAct = pHead + kTransferProtocolHeaderFlagsSize;
```
How do you know pAct is not pointing off the end of the buffer?

crlonxp replied:
Yes. the checking is required

bzbarsky-apple commented on src/wifipaf/WiFiPAFError.cpp:
```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    Copyright (c) 2025 NXP LLC.
```
Again, why?

crlonxp replied:
I will remove it

bzbarsky-apple commented on src/wifipaf/WiFiPAFLayer.h:
```diff
@@ -0,0 +1,204 @@
+
+inline constexpr size_t kCapabilitiesRequestMagicnumLength          = 2;
+inline constexpr size_t kCapabilitiesRequestL2capMtuLength          = 2;
+inline constexpr size_t kCapabilitiesRequestSupportedVersionsLength = 4;
```
Why isn't this computed from NUM_PAFTP_SUPPORTED_PROTOCOL_VERSIONS?  Or at least static_asserted to match it?

crlonxp replied:
I will change to compute it from NUM_PAFTP_SUPPORTED_PROTOCOL_VERSIONS

bzbarsky-apple commented on src/wifipaf/WiFiPAFLayer.h:
```diff
@@ -0,0 +1,204 @@
+     * array elements. Counting up from the zero-index, the first zero-value
+     * specifies the end of the list of supported protocol versions.
+     */
+    uint8_t mSupportedProtocolVersions[(NUM_PAFTP_SUPPORTED_PROTOCOL_VERSIONS / 2) + (NUM_PAFTP_SUPPORTED_PROTOCOL_VERSIONS % 2)];
```
See, here we compute from NUM_PAFTP_SUPPORTED_PROTOCOL_VERSIONS.

The real invariant is that `sizeof(mSupportedProtocolVersions) == kCapabilitiesRequestSupportedVersionsLength`, so why are we not just setting one from the other?  i.e. moving all the constants to be declared in this message class, etc.

bzbarsky-apple commented on src/wifipaf/WiFiPAFTP.cpp:
```diff
@@ -0,0 +1,607 @@
+// Calling convention:
+//   May only be called if data arg is committed for immediate, synchronous subsequent transmission.
+//   Returns false on error. Caller must free data arg on error.
+bool WiFiPAFTP::HandleCharacteristicSend(System::PacketBufferHandle data, bool send_ack)
```
Again, what's with "characteristic"?

crlonxp replied:
It should be the follow-up packets. I will change it

crlonxp commented on examples/chip-tool/commands/pairing/PairingCommand.h:
```diff
@@ -283,7 +289,9 @@ class PairingCommand : public CHIPCommand,
         mComplex_DSTOffsets;
-
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFIPAF
+    chip::Optional<char *> mApFreqStr;
```
Thank you for the suggestion. I will change it.

crlonxp commented on src/include/platform/CHIPDeviceConfig.h:
```diff
@@ -447,6 +447,60 @@
+ *
+ * The default channel of Wi-Fi PAF in 2.4G band (channel#6)
+ */
+#ifndef CHIP_DEVICE_CONFIG_WIFIPAF_24G_DEFAUTL_CHNL
```
Oh, I will re-exam again. Thanks~

chirag-silabs commented on src/include/platform/CHIPDeviceEvent.h:
```diff
@@ -395,6 +397,7 @@ typedef void (*AsyncWorkFunct)(intptr_t arg);
 #include <system/SystemEvent.h>
 #include <system/SystemLayer.h>
 #include <system/SystemPacketBuffer.h>
+#include <wifipaf/WiFiPAFRole.h>
```
Shouldn't this include header also be enclosed in a define? Since this include is not needed for non CHIP_DEVICE_CONFIG_ENABLE_WIFIPAF

crlonxp replied:
I will add it

crlonxp commented on src/include/platform/ConnectivityManager.h:
```diff
@@ -178,16 +178,18 @@ class ConnectivityManager
+    CHIP_ERROR WiFiPAFCancelIncompleteSubscribe();
+    CHIP_ERROR WiFiPAFSend(const WiFiPAF::WiFiPAFSession & TxInfo, System::PacketBufferHandle && msgBuf);
+    WiFiPAF::WiFiPAFLayer * GetWiFiPAF();
+    void WiFiPafSetApFreq(const uint16_t freq);
```
Yes, it's better using the default value. I will change it

crlonxp commented on src/platform/Linux/ConnectivityManagerImpl.cpp:
```diff
@@ -2,6 +2,7 @@
  *
  *    Copyright (c) 2020-2022 Project CHIP Authors
```
Yes, I will change it

crlonxp commented on src/platform/Linux/ConnectivityManagerImpl.cpp:
```diff
@@ -2,6 +2,7 @@
  *
  *    Copyright (c) 2020-2022 Project CHIP Authors
  *    Copyright (c) 2019 Nest Labs, Inc.
+ *    Copyright (c) 2025 NXP
```
Yes, it should be removed

crlonxp commented on src/transport/raw/WiFiPAF.cpp:
```diff
@@ -1,6 +1,6 @@
 /*
  *
- *    Copyright (c) 2024 Project CHIP Authors
+ *    Copyright (c) 2025 Project CHIP Authors
```
Yes, I will change it

crlonxp commented on src/include/platform/internal/GenericPlatformManagerImpl.ipp:
```diff
@@ -152,6 +152,12 @@ void GenericPlatformManagerImpl<ImplClass>::_Shutdown()
 
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFIPAF
+    ChipLogProgress(DeviceLayer, "WiFi-PAF Layer shutdown");
+    WiFiPAF::WiFiPAFLayer::GetWiFiPAFLayer().Shutdown(
```
If the platform shutdowns, we should shutdown all connections. But it may not be required, since it has been done on each individual commissioner.

crlonxp commented on src/platform/Linux/ConnectivityManagerImpl.cpp:
```diff
@@ -101,6 +102,23 @@ namespace DeviceLayer {
+static void StopSignalHandler(int signum)
+{
+    WiFiPAF::WiFiPAFLayer::GetWiFiPAFLayer().Shutdown([](uint32_t id, WiFiPAF::WiFiPafRole role) {
+        switch (role)
```
This may happen while the program is running as either commissioner or commissionee.

arkq commented on src/platform/Linux/ConnectivityManagerImpl.cpp:
```diff
@@ -146,6 +164,16 @@ CHIP_ERROR ConnectivityManagerImpl::_Init()
+    struct sigaction sa = {};
+    sa.sa_handler       = StopSignalHandler;
+    sa.sa_flags         = SA_RESETHAND;
+    sigaction(SIGINT, &sa, nullptr);
```
Please, do not install any signal handlers in the SDK code. Signal handler (if needed) should be installed by the application which uses SDK, not the other way.

https://github.com/project-chip/connectedhomeip/issues/38944
